### PR TITLE
Starred_Messages_Feature/Outreachy

### PR DESCRIPTION
**Pull request by yaya-usman**
Task:  [vector-im/element-web#22453](https://github.com/vector-im/element-web/issues/22453)

Signed-off-by: Yaya Usman [yaya-usman@users.noreply.github.com](mailto:yaya-usman@users.noreply.github.com) 

Type: enhancement

---
Here's what your changelog entry will look like:

## ✨ Features
 * Starred_Messages_Feature/Outreachy ([\#8842](https://github.com/matrix-org/matrix-react-sdk/pull/8842)).

### CODE REVIEW

File: src/components/views/messages/MessageActionBar.tsx
Comment (andybalaam) [inline src/components/views/messages/MessageActionBar.tsx:436]: This comment needs updating to mention the Favourite button.

**Review by andybalaam — CHANGES_REQUESTED** [review]
This looks great. Setting my review to "Request changes" because we agreed that we will wait until the feature actually does something before merging.

File: res/css/views/messages/_MessageActionBar.scss
Comment (andybalaam) [inline res/css/views/messages/_MessageActionBar.scss:122]: We need to give this colour a name and make sure it is themeable. For an example, see https://github.com/matrix-org/matrix-react-sdk/blob/develop/res/themes/light/css/_light.scss#L127 `$menu-border-color`

**Review by andybalaam — CHANGES_REQUESTED** [review]
This looks really good. I don't understand all that useEffect stuff so we'll need someone else to review this, but when you have done all the things I commented on, we should be ready for someone else to review.

The hardest thing I asked for is more tests of the saving code. Feel free to ask me for help on that.

File: src/dispatcher/actions.ts
Comment (yaya-usman) [reply src/dispatcher/actions.ts:326]: Yea, will fix that

File: src/contexts/FavMessageContext.tsx
Comment (yaya-usman) [reply src/contexts/FavMessageContext.tsx:29]: i'll work on that

File: src/components/views/messages/MessageActionBar.tsx
Comment (andybalaam) [reply src/components/views/messages/MessageActionBar.tsx:453]: @yaya-usman I don't think you have updated this comment?

File: test/components/views/messages/MessageActionBar-test.tsx
Comment (andybalaam) [reply test/components/views/messages/MessageActionBar-test.tsx:569]: The new tests address the above comment :+1:

**Review by SimonBrandner — CHANGES_REQUESTED** [review]
Needs to be added to https://github.com/vector-im/element-web/blob/develop/docs/labs.md

File: src/contexts/FavouriteMessageContext.tsx
Comment (andybalaam) [reply src/contexts/FavouriteMessageContext.tsx:49]: Yes, that will be coming soon, but we want to land this as a prototype first.

File: src/contexts/FavouriteMessageContext.tsx
Comment (yaya-usman) [reply src/contexts/FavouriteMessageContext.tsx:48]: I tried to use that once but it fails me, i need to try again today. Thanks

Comment (yaya-usman) [reply src/contexts/FavouriteMessageContext.tsx:48]: @SimonBrandner Can i get a suggestion on how i might use the `MatrixDispatcher ` in my use case since i am using Context, Reducers etc. I tried following the existing approach of how element uses it's dispatch, it's not firing once i click the button that should trigger the dispatch. I hope you understand me. Also is there any advantage of using the `MatrixDispatcher ` over the default React Dispatcher in this case

Comment (SimonBrandner) [reply src/contexts/FavouriteMessageContext.tsx:48]: So I am not sure using a context or `MatrixDispatcher` here makes much sense. IMO, the simplest solution would be to add a custom hook (e.g. `useFavouriteMessages`) which would return an array of the favourited messages and functions to add and remove them

<hr>

Take this with a grain of sand - it might make sense for someone with more experience to take a look; let's not block the PR on this

**Comment by t3chguy** [comment]
This has conflicts and changes requested, please request a review once it is ready for review

File: src/contexts/FavouriteMessageContext.tsx
Comment (andybalaam) [reply src/contexts/FavouriteMessageContext.tsx:48]: @t3chguy please can you review this change, and especially comment on the use of the React dispatcher - is this ok?

File: src/dispatcher/actions.ts
Comment (t3chguy) [inline src/dispatcher/actions.ts:329]: Yeah I second @SimonBrandner - this is actions.ts in the `dispatcher` directory, this is the Flux-based dispatcher. Not the React hooks reducer dispatcher.

**Review by t3chguy — DISMISSED** [review]

File: src/hooks/useFavouriteMessages.ts
Comment (SimonBrandner) [inline src/hooks/useFavouriteMessages.ts:24]: I think you should be able to avoid this using `useLocalStorage()`

**Review by SimonBrandner — COMMENTED** [review]
Looking much better - thank you for the updates!

File: src/components/views/messages/MessageActionBar.tsx
Comment (yaya-usman) [reply src/components/views/messages/MessageActionBar.tsx:249]: Yes exactly, it definely would. Let me change that. Thanks

Comment (yaya-usman) [reply src/components/views/messages/MessageActionBar.tsx:249]: Ohh wait, you mean instead of ` handle click` it should be `toggleFavourite(eventId)` right?

File: src/components/views/messages/MessageActionBar.tsx
Comment (yaya-usman) [reply src/components/views/messages/MessageActionBar.tsx:254]: is there a method like that?

File: src/components/views/messages/MessageActionBar.tsx
Comment (yaya-usman) [reply src/components/views/messages/MessageActionBar.tsx:243]: Got it

File: src/hooks/useFavouriteMessages.ts
Comment (yaya-usman) [reply src/hooks/useFavouriteMessages.ts:24]: you mean instead of `localStorage.getItem` i should use the `useLocalStorage`  hook?

File: src/components/views/messages/MessageActionBar.tsx
Comment (SimonBrandner) [reply src/components/views/messages/MessageActionBar.tsx:254]: You can import it: `import classNames from 'classnames';`

File: src/hooks/useFavouriteMessages.ts
Comment (yaya-usman) [reply src/hooks/useFavouriteMessages.ts:24]: Also @SimonBrandner  please, i don't know why the i18n is failing, i tried to check if it's from any of my file but apparently not

Comment (SimonBrandner) [reply src/hooks/useFavouriteMessages.ts:24]: > Also @SimonBrandner please, i don't know why the i18n is failing, i tried to check if it's from any of my file but apparently not

Have you run `yarn i18n`

Comment (yaya-usman) [reply src/hooks/useFavouriteMessages.ts:24]: No i didn't run that, let me try it

Comment (yaya-usman) [reply src/hooks/useFavouriteMessages.ts:24]: useLocal

Comment (yaya-usman) [reply src/hooks/useFavouriteMessages.ts:24]: apparently, the hook itself has never been used before when i tried to search it's use case in the project, i guess due to this reason. Also, i am only using local storage temporarily, the code gonna change later on when i move to accountdata

Comment (SimonBrandner) [reply src/hooks/useFavouriteMessages.ts:24]: This looks odd and probably shouldn't happen... @t3chguy, do you have any idea what is happening here?

Comment (SimonBrandner) [reply src/hooks/useFavouriteMessages.ts:24]: Ok, so it turns out `useLocalStorage` can't be used here as the localStorage value can be altered from multiple places - let me think about a better solution

**Review by andybalaam — APPROVED** [review]
Looks good!

File: src/hooks/useFavouriteMessages.ts
Comment (SimonBrandner) [inline src/hooks/useFavouriteMessages.ts:37]: Could you please add a comment about why is this necessary

**Review by SimonBrandner — APPROVED** [review]
After further thought even though the way `localStorage` is used here isn't ideal, we can ignore it as it will soon be replaced

LGTM - thank you!

File: src/dispatcher/actions.ts
Comment (andybalaam) [reply src/dispatcher/actions.ts]: @t3chguy please can you remove your change request since Usman solved this a different way and has a tick from @SimonBrandner and me?

**Review by SimonBrandner — APPROVED** [review]
Thank you!

(will let Andy press the button; also, please fix the merge conflicts)

**Comment by yaya-usman** [comment]
> Thank you!
> 
> (will let Andy press the button; also, please fix the merge conflicts)

Thanks but i don't know why i couldn't, it's weird

**Comment by yaya-usman** [comment]
conflict

**Comment by SimonBrandner** [comment]
> conflict

You need to manually resolve them in your IDE/code editor

**Comment by yaya-usman** [comment]
> conflict

This is what i meant, i don't know why i couldn't click the button, also i don't have any conflicts in my local version. So it's weird

**Comment by SimonBrandner** [comment]
> > Thank you!
> > (will let Andy press the button; also, please fix the merge conflicts)
> 
> @SimonBrandner Andy asked me to click merge if i made the changes you requested and the checks passes. I hope it's ok by you. Also Thanks for the review!

Sounds good :+1:

**Pull request merged**
